Opening the GIT checkout bundle fails due to non-empty target folder

Recent occurrence:

https://dev.azure.com/dnceng/public/_build/results?buildId=464092&view=logs&j=03e59d4a-1319-5b6e-0c0f-bf32e6fef540&t=5aa541f1-82f6-5969-3e66-2cde7c445b54

(Linux x64 debug):

<pre>
/usr/bin/bash --noprofile --norc /__w/_temp/76490d1e-57f6-45d9-a4f8-dec0a0149235.sh
+ git clone /__w/3/Checkout_bundle/Checkout.bundle .
fatal: destination path '.' already exists and is not an empty directory.
##[error]Bash exited with code '128'.
Finishing: Clone the repository from Checkout.bundle
</pre>

Please note that the failure looks similar to #736 but the root cause is different. AFAIK we're not always cleaning up the containers we're running Linux tests in, @janvorli might be able to provide more detail. We should probably explicitly do the cleanup prior to opening the bundle.

On Linux, this should be a piece of cake with rm -r. On Windows, an additional wrinkle is that, as directory deletion is inherently asynchronous, rmdir /s /q non-deterministically fails if a reference to a folder is kept somewhere. I fought this at length in .NET Native infra and I ended up with a custom C# tool, RmDirHardLink, that was among others solving the problem using a retry mechanism.

--- a/eng/pipelines/common/clone-checkout-bundle-step.yml
+++ b/eng/pipelines/common/clone-checkout-bundle-step.yml
@@ -1,6 +1,5 @@
 steps:
 - checkout: none
-  clean: true
 
 - download: current
   artifact: Checkout_bundle

--- a/eng/pipelines/common/global-build-job.yml
+++ b/eng/pipelines/common/global-build-job.yml
@@ -18,6 +18,8 @@ jobs:
     container: ${{ parameters.container }}
     timeoutInMinutes: ${{ parameters.timeoutInMinutes }}
     variables: ${{ parameters.variables }}
+    workspace:
+      clean: all
 
     steps:
     - ${{ if eq(parameters.osGroup, 'OSX') }}:

